File: windows_app/savestories_windows/worker_client.py
from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
from dataclasses import asdict
from typing import Sequence

from .app_paths import AppPaths
from .models import WorkerBatchResult, WorkerCounts, WorkerItem, WorkerRequest, WorkerResponse, WorkerRuntime


class WorkerClient:
    def __init__(self) -> None:
        self.current_process: subprocess.Popen[bytes] | None = None
        self.login_process: subprocess.Popen[bytes] | None = None

    def stop_current_process(self) -> None:
        process = self.current_process
        if process is None or process.poll() is not None:
            return
        process.terminate()

    def start_detached_login(self, request: WorkerRequest) -> None:
        AppPaths.ensure_directories()
        command, runtime = self.resolve_command(request)

        environment = os.environ.copy()
        environment["SAVESTORIES_APP_SUPPORT"] = str(AppPaths.application_support())
        environment["SAVESTORIES_BROWSER_PROFILE"] = str(AppPaths.browser_profile())
        environment["SAVESTORIES_MANIFESTS"] = str(AppPaths.manifests_directory())
        environment["SAVESTORIES_PLAYWRIGHT_BROWSERS"] = str(AppPaths.playwright_browsers())
        environment["SAVESTORIES_DEFAULT_DOWNLOADS"] = str(AppPaths.default_downloads())
        environment["SAVESTORIES_LOGS"] = str(AppPaths.logs_directory())
        environment["SAVESTORIES_WORKER_RUNTIME"] = runtime

        popen_options = self._windows_popen_options(detached=True)

        process = subprocess.Popen(
            command,
            stdin=subprocess.PIPE,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            env=environment,
            close_fds=True,
            **popen_options,
        )
        self.login_process = process
        assert process.stdin is not None
        process.stdin.write((json.dumps(asdict(request)) + "\n").encode("utf-8"))
        process.stdin.flush()
        process.stdin.close()

    def run(self, request: WorkerRequest) -> WorkerResponse:
        AppPaths.ensure_directories()
        command, runtime = self.resolve_command(request)

        environment = os.environ.copy()
        environment["SAVESTORIES_APP_SUPPORT"] = str(AppPaths.application_support())
        environment["SAVESTORIES_BROWSER_PROFILE"] = str(AppPaths.browser_profile())
        environment["SAVESTORIES_MANIFESTS"] = str(AppPaths.manifests_directory())
        environment["SAVESTORIES_PLAYWRIGHT_BROWSERS"] = str(AppPaths.playwright_browsers())
        environment["SAVESTORIES_DEFAULT_DOWNLOADS"] = str(AppPaths.default_downloads())
        environment["SAVESTORIES_LOGS"] = str(AppPaths.logs_directory())
        environment["SAVESTORIES_WORKER_RUNTIME"] = runtime

        popen_options = self._windows_popen_options(detached=False)

        process = subprocess.Popen(
            command,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            env=environment,
            close_fds=True,
            **popen_options,
        )
        self.current_process = process
        stdout_data, stderr_data = process.communicate((json.dumps(asdict(request)) + "\n").encode("utf-8"))
        self.current_process = None

        if process.returncode != 0 and not stdout_data:
            stderr_text = stderr_data.decode("utf-8", errors="replace").strip()
            if not stderr_text:
                stderr_text = f"Worker process failed with exit status {process.returncode}."
            return WorkerResponse.process_failure(stderr_text)

        if not stdout_data:
            stderr_text = stderr_data.decode("utf-8", errors="replace").strip() or "Worker returned no output."
            return WorkerResponse.process_failure(stderr_text)

        stdout_text = stdout_data.decode("utf-8", errors="replace")
        try:
            payload, stdout_tail = self._load_first_json_object(stdout_text)
        except Exception:
            return WorkerResponse.process_failure(f"Worker returned invalid JSON.\n{stdout_text}")

        items = [WorkerItem(**item) for item in payload.get("items", [])]
        counts_payload = payload.get("counts")
        counts = WorkerCounts(**counts_payload) if isinstance(counts_payload, dict) else None
        batch_results = [
            WorkerBatchResult(**item)
            for item in payload.get("batchResults", [])
            if isinstance(item, dict)
        ]
        runtime_payload = payload.get("runtime")
        runtime = WorkerRuntime(**runtime_payload) if isinstance(runtime_payload, dict) else None
        return WorkerResponse(
            ok=payload.get("ok", False),
            status=payload.get("status", "unknown"),
            message=payload.get("message", ""),
            protocolVersion=payload.get("protocolVersion"),
            data=payload.get("data", {}) or {},
            counts=counts,
            batchResults=batch_results,
            runtime=runtime,
            diagnostics=payload.get("diagnostics", {}) or {},
            items=items,
            logs=(payload.get("logs", []) or []) + ([f"worker_stdout_tail={stdout_tail}"] if stdout_tail else []),
        )

    @staticmethod
    def _load_first_json_object(stdout_text: str) -> tuple[dict, str]:
        decoder = json.JSONDecoder()
        start = stdout_text.find("{")
        if start < 0:
            raise ValueError("Worker stdout does not contain a JSON object.")
        payload, end = decoder.raw_decode(stdout_text[start:])
        if not isinstance(payload, dict):
            raise ValueError("Worker JSON response must be an object.")
        return payload, stdout_text[start + end :].strip()

    @staticmethod
    def resolve_command(request: WorkerRequest | None = None) -> tuple[Sequence[str], str]:
        try:
            script = AppPaths.node_worker_script()
            return ([str(AppPaths.node_executable()), str(script)], "node")
        except FileNotFoundError:
            pass

        try:
            script = AppPaths.node_worker_script()
            node = shutil.which("node")
            if node:
                return ([node, str(script)], "node")
        except FileNotFoundError:
            pass

        if getattr(sys, "frozen", False):
            raise RuntimeError("Не удалось найти встроенный Node runtime. Переустанови приложение.")

        raise RuntimeError("Node runtime не найден. Подготовь движок в настройках приложения.")

    @staticmethod
    def resolve_python_command() -> Sequence[str]:
        if getattr(sys, "frozen", False):
            return [sys.executable, "--worker-bridge"]

        candidate = AppPaths.worker_python()
        if candidate.exists():
            return [str(candidate)]

        py_launcher = shutil.which("py")
        if py_launcher:
            return [py_launcher, "-3"]

        python = shutil.which("python")
        if python:
            return [python]

        raise RuntimeError("Python 3 не найден. Установи Python 3.13+ и затем выполни настройку движка.")

    @staticmethod
    def _windows_popen_options(*, detached: bool) -> dict:
        if os.name != "nt":
            return {}

        creationflags = subprocess.CREATE_NO_WINDOW
        if detached:
            creationflags |= subprocess.DETACHED_PROCESS | subprocess.CREATE_NEW_PROCESS_GROUP

        startupinfo = subprocess.STARTUPINFO()
        startupinfo.dwFlags |= subprocess.STARTF_USESHOWWINDOW
        startupinfo.wShowWindow = 0
        return {
            "creationflags": creationflags,
            "startupinfo": startupinfo,
        }

File: Sources/SaveStories/Services/WorkerClient.swift
import Foundation

@MainActor
final class WorkerClient {
    struct LaunchConfiguration {
        let executable: URL
        let arguments: [String]
        let runtime: String
    }

    private final class CompletionGate: @unchecked Sendable {
        private let lock = NSLock()
        private var completed = false
        private var inputFailure: String?

        func claim() -> Bool {
            lock.lock()
            defer { lock.unlock() }
            guard !completed else { return false }
            completed = true
            return true
        }

        func recordInputFailure(_ message: String) {
            lock.lock()
            defer { lock.unlock() }
            inputFailure = message
        }

        func inputFailureMessage() -> String? {
            lock.lock()
            defer { lock.unlock() }
            return inputFailure
        }
    }

    private final class PipeCollector: @unchecked Sendable {
        private let fileHandle: FileHandle
        private let onLine: (@Sendable (String) -> Void)?
        private let lock = NSLock()
        private let finished = DispatchSemaphore(value: 0)
        private var collectedData = Data()
        private var readerThread: Thread?

        init(fileHandle: FileHandle, onLine: (@Sendable (String) -> Void)? = nil) {
            self.fileHandle = fileHandle
            self.onLine = onLine
        }

        func start() {
            guard readerThread == nil else { return }

            let thread = Thread { [fileHandle, onLine] in
                var completeData = Data()
                var partialLine = Data()

                while true {
                    let chunk = fileHandle.availableData
                    if chunk.isEmpty {
                        break
                    }

                    completeData.append(chunk)

                    guard let onLine else { continue }
                    partialLine.append(chunk)

                    while let newlineIndex = partialLine.firstIndex(of: 0x0A) {
                        let lineData = partialLine.prefix(upTo: newlineIndex)
                        partialLine.removeSubrange(...newlineIndex)
                        if let line = String(data: lineData, encoding: .utf8)?
                            .trimmingCharacters(in: .whitespacesAndNewlines),
                           !line.isEmpty
                        {
                            onLine(line)
                        }
                    }
                }

                if let onLine,
                   let trailing = String(data: partialLine, encoding: .utf8)?
                    .trimmingCharacters(in: .whitespacesAndNewlines),
                   !trailing.isEmpty
                {
                    onLine(trailing)
                }

                self.lock.lock()
                self.collectedData = completeData
                self.lock.unlock()
                self.finished.signal()
            }
            thread.name = "SaveMe.PipeCollector"
            thread.start()
            readerThread = thread
        }

        func finish(timeout: TimeInterval = 5.0) -> Data {
            _ = finished.wait(timeout: .now() + timeout)
            lock.lock()
            defer { lock.unlock() }
            return collectedData
        }
    }

    enum WorkerClientError: LocalizedError {
        case workerScriptNotFound
        case processLaunchFailed(String)
        case invalidWorkerResponse(byteCount: Int)

        var errorDescription: String? {
            switch self {
            case .workerScriptNotFound:
                "Worker script was not found in the package resources."
            case .processLaunchFailed(let message):
                message
            case .invalidWorkerResponse(let byteCount):
                "Worker вернул некорректный JSON (\(byteCount) байт). Попробуйте повторить операцию или подготовить среду заново."
            }
        }
    }

    private var currentProcess: Process?
    private var userInitiatedStop = false
    private let prepareEnvironment: () throws -> Void
    private let launchConfigurationOverride: LaunchConfiguration?

    init(
        prepareEnvironment: @escaping () throws -> Void = {
            try AppPaths.ensureDirectories()
            try AppPaths.synchronizeBundledNodeWorkerSources()
        },
        launchConfiguration: LaunchConfiguration? = nil
    ) {
        self.prepareEnvironment = prepareEnvironment
        self.launchConfigurationOverride = launchConfiguration
    }

    func run(_ request: WorkerRequest, onProgress: (@Sendable (String) -> Void)? = nil) async -> WorkerResponse {
        guard currentProcess == nil else {
            return .processFailure(message: "Worker уже выполняет другую операцию. Дождитесь её завершения.")
        }
        guard !Task.isCancelled else {
            return .cancelled(message: "Операция отменена.")
        }
        defer {
            currentProcess = nil
            userInitiatedStop = false
        }

        do {
            let response = try await execute(request, onProgress: onProgress)
            if userInitiatedStop || Task.isCancelled {
                return .cancelled(message: "Загрузка остановлена пользователем.")
            }
            return response
        } catch {
            if userInitiatedStop || Task.isCancelled {
                return .cancelled(message: "Загрузка остановлена пользователем.")
            }
            return .processFailure(message: error.localizedDescription)
        }
    }

    func stopCurrentProcess() {
        guard let currentProcess, currentProcess.isRunning else { return }
        userInitiatedStop = true
        currentProcess.terminate()
    }

    private func execute(_ request: WorkerRequest, onProgress: (@Sendable (String) -> Void)? = nil) async throws -> WorkerResponse {
        try prepareEnvironment()
        var encodedRequest = try JSONEncoder().encode(request)
        encodedRequest.append(0x0A)

        let process = Process()
        let stdinPipe = Pipe()
        let stdoutPipe = Pipe()
        let stderrPipe = Pipe()
        let stdoutCollector = PipeCollector(fileHandle: stdoutPipe.fileHandleForReading)
        let stderrCollector = PipeCollector(fileHandle: stderrPipe.fileHandleForReading, onLine: onProgress)

        let launch = try launchConfigurationOverride ?? workerLaunchConfiguration()
        process.executableURL = launch.executable
        process.arguments = launch.arguments

        process.standardInput = stdinPipe
        process.standardOutput = stdoutPipe
        process.standardError = stderrPipe

        var environment = ProcessInfo.processInfo.environment
        environment["SAVESTORIES_APP_SUPPORT"] = AppPaths.applicationSupport.path
        environment["SAVESTORIES_BROWSER_PROFILE"] = AppPaths.browserProfile.path
        environment["SAVESTORIES_MANIFESTS"] = AppPaths.manifestsDirectory.path
        environment["SAVESTORIES_PLAYWRIGHT_BROWSERS"] = AppPaths.playwrightBrowsers.path
        environment["SAVESTORIES_DEFAULT_DOWNLOADS"] = AppPaths.defaultDownloads.path
        environment["SAVESTORIES_LOGS"] = AppPaths.logsDirectory.path
        environment["SAVESTORIES_WORKER_RUNTIME"] = launch.runtime
        if let mediaMuxer = AppPaths.bundledMediaMuxerExecutable {
            environment["SAVEME_MEDIA_MUXER"] = mediaMuxer.path
        }
        if let bundledFrameworks = AppPaths.bundledFrameworksDirectory {
            environment["DYLD_FRAMEWORK_PATH"] = bundledFrameworks.path
        }
        if launch.runtime == "python", let bundledPythonHome = AppPaths.bundledPythonHome {
            environment["PYTHONHOME"] = bundledPythonHome.path
            environment["PYTHONNOUSERSITE"] = "1"
        }
        if launch.runtime == "python", let bundledSitePackages = AppPaths.bundledSitePackages {
            environment["PYTHONPATH"] = bundledSitePackages.path
        }
        process.environment = environment
        currentProcess = process

        let responseData: Data = try await withTaskCancellationHandler {
            try Task.checkCancellation()
            return try await withCheckedThrowingContinuation { continuation in
            let completionGate = CompletionGate()

            process.terminationHandler = { process in
                let stdoutData = stdoutCollector.finish()
                let stderrData = stderrCollector.finish()
                guard completionGate.claim() else { return }
                if let inputFailure = completionGate.inputFailureMessage() {
                    continuation.resume(throwing: WorkerClientError.processLaunchFailed(inputFailure))
                    return
                }

                if process.terminationStatus != 0 && stdoutData.isEmpty {
                    let stderrText = String(data: stderrData, encoding: .utf8)?
                        .trimmingCharacters(in: .whitespacesAndNewlines)
                    let reason: String
                    switch process.terminationReason {
                    case .exit:
                        reason = "exit status \(process.terminationStatus)"
                    case .uncaughtSignal:
                        reason = "signal \(process.terminationStatus)"
                    @unknown default:
                        reason = "unknown termination \(process.terminationStatus)"
                    }

                    let message: String
                    if let stderrText, !stderrText.isEmpty {
                        message = "\(stderrText)\n[\(reason)]"
                    } else {
                        message = "Worker process failed with \(reason)."
                    }
                    continuation.resume(throwing: WorkerClientError.processLaunchFailed(message))
                    return
                }

                if stdoutData.isEmpty {
                    let stderrText = String(data: stderrData, encoding: .utf8)?
                        .trimmingCharacters(in: .whitespacesAndNewlines)
                    let message: String
                    if let stderrText, !stderrText.isEmpty {
                        message = stderrText
                    } else {
                        message = "Worker returned no output."
                    }
                    continuation.resume(throwing: WorkerClientError.processLaunchFailed(message))
                    return
                }

                continuation.resume(returning: stdoutData)
            }

            do {
                try process.run()
                stdoutCollector.start()
                stderrCollector.start()
            } catch {
                guard completionGate.claim() else { return }
                continuation.resume(
                    throwing: WorkerClientError.processLaunchFailed("Failed to launch worker: \(error.localizedDescription)")
                )
                return
            }

            do {
                try stdinPipe.fileHandleForWriting.write(contentsOf: encodedRequest)
                try stdinPipe.fileHandleForWriting.close()
            } catch {
                completionGate.recordInputFailure("Failed to send request to worker: \(error.localizedDescription)")
                try? stdinPipe.fileHandleForWriting.close()
                if process.isRunning {
                    process.terminate()
                }
            }
            }
        } onCancel: {
            Task { @MainActor [weak self] in
                guard self?.currentProcess === process else { return }
                self?.stopCurrentProcess()
            }
        }

        do {
            return try JSONDecoder().decode(WorkerResponse.self, from: responseData)
        } catch {
            throw WorkerClientError.invalidWorkerResponse(byteCount: responseData.count)
        }
    }

    private func workerLaunchConfiguration() throws -> LaunchConfiguration {
        if let nodeScript = nodeWorkerScriptURL(),
           let installedNode = AppPaths.installedNodeExecutable {
            return LaunchConfiguration(executable: installedNode, arguments: [nodeScript.path], runtime: "node")
        }

        if let nodeScript = nodeWorkerScriptURL(),
           let bundledNode = AppPaths.bundledNodeExecutable {
            return LaunchConfiguration(executable: bundledNode, arguments: [nodeScript.path], runtime: "node")
        }

        if let nodeScript = nodeWorkerScriptURL(),
           let nodeExecutable = locateExecutable(named: "node") {
            return LaunchConfiguration(executable: nodeExecutable, arguments: [nodeScript.path], runtime: "node")
        }

        throw WorkerClientError.processLaunchFailed("Node runtime не найден. Подготовьте среду воркера в настройках приложения.")
    }

    private func nodeWorkerScriptURL() -> URL? {
        let installedScript = AppPaths.installedNodeWorkerRoot?
            .appendingPathComponent("bridge.mjs", isDirectory: false)
        let fallbackInstalledScript = AppPaths.workerRoot
            .appendingPathComponent("bridge.mjs", isDirectory: false)
        let candidates = [
            installedScript,
            Bundle.main.sharedSupportURL?
                .appendingPathComponent("node_worker", isDirectory: true)
                .appendingPathComponent("bridge.mjs", isDirectory: false),
            Bundle.main.bundleURL
                .appendingPathComponent("Contents", isDirectory: true)
                .appendingPathComponent("SharedSupport", isDirectory: true)
                .appendingPathComponent("node_worker", isDirectory: true)
                .appendingPathComponent("bridge.mjs", isDirectory: false),
            fallbackInstalledScript,
            URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
                .appendingPathComponent("node_worker", isDirectory: true)
                .appendingPathComponent("bridge.mjs", isDirectory: false),
        ]

        return candidates
            .compactMap { $0 }
            .first(where: { FileManager.default.fileExists(atPath: $0.path) })
    }

    private func pythonWorkerScriptURL() throws -> URL {
        if let bundled = bundledResourceURL(relativePath: "worker/bridge.py"),
           FileManager.default.fileExists(atPath: bundled.path) {
            return bundled
        }

        let fallbackCandidates = [
            URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
                .appendingPathComponent("Sources", isDirectory: true)
                .appendingPathComponent("SaveMe", isDirectory: true)
                .appendingPathComponent("Resources", isDirectory: true)
                .appendingPathComponent("worker", isDirectory: true)
                .appendingPathComponent("bridge.py", isDirectory: false),
            URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
                .appendingPathComponent("Sources", isDirectory: true)
                .appendingPathComponent("SaveStories", isDirectory: true)
                .appendingPathComponent("Resources", isDirectory: true)
                .appendingPathComponent("worker", isDirectory: true)
                .appendingPathComponent("bridge.py", isDirectory: false),
        ]

        if let fallback = fallbackCandidates.first(where: { FileManager.default.fileExists(atPath: $0.path) }) {
            return fallback
        }

        throw WorkerClientError.workerScriptNotFound
    }

    private func locateExecutable(named executable: String) -> URL? {
        let pathVariable = ProcessInfo.processInfo.environment["PATH"] ?? ""
        for component in pathVariable.split(separator: ":") {
            let candidate = URL(fileURLWithPath: String(component), isDirectory: true)
                .appendingPathComponent(executable, isDirectory: false)
            if FileManager.default.isExecutableFile(atPath: candidate.path) {
                return candidate
            }
        }
        return nil
    }

    private func bundledResourceURL(relativePath: String) -> URL? {
        let fileManager = FileManager.default
        let candidates = [
            Bundle.main.resourceURL?
                .appendingPathComponent(AppPaths.resourceBundleName, isDirectory: true)
                .appendingPathComponent("Contents/Resources", isDirectory: true)
                .appendingPathComponent(relativePath, isDirectory: false),
            Bundle.main.resourceURL?
                .appendingPathComponent(AppPaths.resourceBundleName, isDirectory: true)
                .appendingPathComponent(relativePath, isDirectory: false),
            Bundle.main.bundleURL
                .appendingPathComponent("Contents", isDirectory: true)
                .appendingPathComponent("Resources", isDirectory: true)
                .appendingPathComponent(AppPaths.resourceBundleName, isDirectory: true)
                .appendingPathComponent(relativePath, isDirectory: false),
        ]

        return candidates
            .compactMap { $0 }
            .first(where: { fileManager.fileExists(atPath: $0.path) })
    }
}
